### DISCUSSION TITLE
Syncing nested objects in Zustand

### DISCUSSION THREAD

**Discussion by therealtimhawkins** [liveblocks/liveblocks#1614] [Help] [answered]
I am using Liveblocks, along with Zustand, to sync a Reactflow chart. 

The issue that I'm having is that I have meta data stored in a "metadata" object on the Reactflow nodes, and when I update this metadata the changes aren't saved to Liveblocks. Whenever a parameter is changed that isn't nested, however, it is updated fine.

Is it possible to have nested LiveLists and LiveObjects or am I misunderstanding how they work? 

Apologies if this is similar to the previous [question about nested objects](https://github.com/liveblocks/liveblocks/discussions/1584). I'm not sure that answers my question though as I'm using `@liveblocks/zustand` rather than the Live* structures directly?

Thanks! Great work btw, really great product 🙌

**Comment by nvie** [comment]
> The issue that I'm having is that I have meta data stored in a "metadata" object on the Reactflow nodes, and when I update this metadata the changes aren't saved to Liveblocks. Whenever a parameter is changed that isn't nested, however, it is updated fine.

Are the Reactflow nodes stored in your Zustand store, and are you using `set()` to update them? Any call to `set()` which changes one of the top-level properties defined in your `storageMapping` config should get synchronized with Liveblocks automatically.

Maybe we can provide a more direct answer to you if you share your Liveblocks config file, and the code that mutates your Reactflow metadata. Thanks!

**Reply by therealtimhawkins** [reply]
Yes they are stored in Zustand and I use `set()` to update them.

Here is the action to update the node data:

```
      updateNode: (
        id: string,
        data: Partial<MaterialNodeDataType | CategoryNodeDataType>
      ) => {
        const nodes = get().nodes.map((node: Node) => {
          if (node.id === id) {
            node.data = { ...node.data, ...data };
          }
          return node;
        });
        set({
          nodes,
        });
      },
```

I'm not sure which Liveblocks config file you mean? At the moment I'm just using:

```
const client = createClient({
  publicApiKey:
    "secret-key",
});
```

and `@liveblocks/zustand` in my store.

Thanks, I hope this helps!

**Reply by nvie** [reply, marked as answer]
I think the most likely reason for the update not propagating through here is that you're not changing the reference in an immutable way, which is what Zustand (and our `@liveblocks/zustand` integration) expects. Does it work when you change the line as follows?

```diff
         const nodes = get().nodes.map((node: Node) => {
           if (node.id === id) {
-             node.data = { ...node.data, ...data };
+             const newData = { ...node.data, ...data };
+             node = { ...node, data: newData };
           }
           return node;
         });
```

The reason is that if from the outside it's still the same object reference, both Zustand and Liveblocks will assume the object hasn't changed. (Neither will do a deep comparison, for performance reasons. Here is the [relevant section](https://docs.pmnd.rs/zustand/guides/updating-state#deeply-nested-object) from the Zustand docs.)

Hope this helps to resolve it for you!
